### DISCUSSION TITLE
zod on readonly field keep getting invalid

### DISCUSSION THREAD

**Discussion by vatoer** [colinhacks/zod#3215] [Q&A] [answered]
hi, 

I have an Input which will be filled by a calendar picker, 

here is my simplified input component 

**will always throw error invalid date**

```tsx
<input
    readOnly
      value={
        date
          ? format(date, "yyyy-MM-dd", { locale: props.locale ?? id })
          : ""
      }
      type={"text"}
      id={name}
      {...register(name)}
/>
```

this will validate as expected but will not update the value picked on Calendar
```tsx
<input
         type={"text"}
      id={name}
      {...register(name)}
/>
```

and my schema 

```ts
export const genericDateSchema = z.coerce.date();

export const spriSchema = z
  .object({
    birthDate: genericTanggalSchema,
      })
```

what is the best approach for this? 
thanks in advanced

**Comment by JacobWeisenburger** [comment]
Please send a full reproducible code example.

**Reply by vatoer** [reply]
# sandbox 
here is my sandbox

https://codesandbox.io/p/github/vatoer/sandbox-zod-playgroud/main

here is the github

[git@github.com:vatoer/sandbox-zod-playgroud.git](https://github.com/vatoer/sandbox-zod-playgroud.git)

## how to reproduce 

1. browse the page 
2. fill name, birth date, journey date, pick date from datepicker
3. click submit

![image](https://github.com/colinhacks/zod/assets/8667392/316557b3-ff5a-4e24-9ec8-c80c53cdf4f9)


## expected result 
no error, console show the data

## current result 
on first submit click, 
-birth date , journey date give error date invalid

on second  submit click, 
-journey date give error date invalid

on 3rd click 
-no error, console show the data

## additional info

I have to click (n) times as much as my datepicker component

**Comment by JacobWeisenburger** [comment, marked as answer]
you're `handleSelect` function inside of `InputDatePicker` never actually sets the value of that input on the form. You are setting the value of the input thru html, but you aren't setting it on the react hook form. This has nothing to do with Zod, so if you need further help, feel free to make an issue/discussion on `react-hook-form`'s github.

I found this in their docs, perhaps it could help. https://react-hook-form.com/docs/usecontroller/controller

**Reply by vatoer** [reply]
thanks,  

as your suggestion, fixed by using setValue from useForm when update the input value, 

```
const {
    register,
    setValue,
    handleSubmit,
    formState: { errors },
  } = useForm<FormData>({
    resolver: zodResolver(spriSchema),
  });
  ```
  
  and then 
  ```
  const handleSelect: SelectSingleEventHandler = (newDate) => {
    setDate(newDate ?? date);
    setIsPopoverOpen(false);
    if (newDate) {
      const newDateStr = format(newDate, "PPP", {
        locale: props.locale ?? id,
      });
      setValue(name, newDateStr);
    }
  };
  ```

updated the code on github with the working one for future me

**Comment by vatoer** [comment]


**Reply by JacobWeisenburger** [reply]
You might have to use `trigger` from `react-hook-form`. I'm not an expert for that. But I'm happy to answer zod related questions in more depth.

**Reply by vatoer** [reply]
many thanks for the hint. really appreciate your help. got it, so it still related on `react-hook-form`

https://www.react-hook-form.com/api/useform/setvalue/ 

have option shouldValidate

````ts
      setValue(name, newDateStr, { shouldValidate: true });
````

fixed
